### PR TITLE
add: dyo cli

### PR DESCRIPTION
**Pull request by c3ppc3pp**
Dyo command line utility, right now it generates a docker compose file with defined parameters.

Work in Progress

### CODE REVIEW

File: dyo/dyo.go
Comment (polaroi8d) [inline dyo/dyo.go:11]: If I run with a random command, it will exit with a random(? for me) status. It would be nice if an unknown command given by the user, show the help page. Same with the flags.

File: dyo/dyo.go
Comment (polaroi8d) [inline dyo/dyo.go:13]: Naming is still cli, imho.

File: dyo/dyo.go
Comment (polaroi8d) [inline dyo/dyo.go:8]: Any reason why are you picked urfave instead of cobra?

File: dyo/configuration.go
Comment (polaroi8d) [inline dyo/configuration.go:265]: Unsued log.

File: dyo/configuration.go
Comment (polaroi8d) [inline dyo/configuration.go:205]: We can outsource these types of templates into files, IMHO. As we made in node script generation.

File: dyo/configuration.go
Comment (polaroi8d) [inline dyo/configuration.go:203]: Use CamelCase naming convention.

File: web/crux/Dockerfile
Comment (polaroi8d) [inline web/crux/Dockerfile:13]: Could you name the Stage #1 to, please?

File: dyo/dyo.go
Comment (polaroi8d) [inline dyo/dyo.go:61]: Minor but we are using `cruxui` in `crux-ui` way.

**Review by polaroi8d — CHANGES_REQUESTED** [review]
Right now when I run init, I got an almost empty compose file, with a full commented settings.yaml, is this the desired state? 

Missing the inputs, when running init, which service I would like to bootstrap, and what will be the data (default or the user would like to overwrite), and so on.

File: dyo/dyo.go
Comment (c3ppc3pp) [reply dyo/dyo.go:8]: Seemed much more straightforward, and simple compared to cobra. Cobra is a good candidate if you want to build a spaceship, I'd like to take things more lightly if it's not necessary.

File: dyo/config.go
Comment (c3ppc3pp) [reply dyo/config.go]: Fixed, now every file is available under the cli directory.

File: dyo/configuration.go
Comment (c3ppc3pp) [reply dyo/configuration.go]: Removed. Thanks!

File: dyo/dyo.go
Comment (c3ppc3pp) [reply dyo/dyo.go]: Will address this soon!

File: dyo/configuration.go
Comment (c3ppc3pp) [reply dyo/configuration.go]: Removed.

File: dyo/configuration.go
Comment (c3ppc3pp) [reply dyo/configuration.go]: Indeed, then embed it so the binary can be standalone.

File: dyo/dyo.go
Comment (c3ppc3pp) [reply dyo/dyo.go]: Renamed the const! It was the result of an earlier attempt, also go can eff themselves with their "enums". :-1: 
Good catch! Thanks!

**Comment by c3ppc3pp** [comment]
> Right now when I run init, I got an almost empty compose file, with a full commented settings.yaml, is this the desired state?
> 
> Missing the inputs, when running init, which service I would like to bootstrap, and what will be the data (default or the user would like to overwrite), and so on.

`go run . init ` should set you up now. I missed an else branch. Thanks!

File: agent/internal/grpc/grpc.go
Comment (m8vago) [reply agent/internal/grpc/grpc.go:414]: @nandor-magyar ?

File: dyo/dyo.go
Comment (nandor-magyar) [reply dyo/dyo.go]: In the repository, the directory/package should be cli, but otherwise the binary and the alias can be **dyo**.

File: agent/internal/grpc/grpc.go
Comment (nandor-magyar) [reply agent/internal/grpc/grpc.go:414]: Remove it.

File: dyo/dyo.go
Comment (polaroi8d) [reply dyo/dyo.go]: I can agree with that.

File: dyo/dyo.go
Comment (c3ppc3pp) [reply dyo/dyo.go]: fixed

**Comment by c3ppc3pp** [comment]
If anyone want to try it out, feel free!

File: cli/containergen.go
Comment (polaroi8d) [inline cli/containergen.go:1]: We are using snake_case for go files.

File: cli/runner.go
Comment (polaroi8d) [inline cli/runner.go:166]: `error should be returned as the last argument ST1008`

File: cli/runner.go
Comment (polaroi8d) [inline cli/runner.go:62]: `error should be returned as the last argument ST1008`

File: cli/runner.go
Comment (polaroi8d) [inline cli/runner.go:189]: err never used, dead code.

File: cli/runner.go
Comment (polaroi8d) [inline cli/runner.go:195]: Same as above.

File: cli/runner.go
Comment (polaroi8d) [inline cli/runner.go:256]: Error strings should not be capitalized.

File: cli/runner.go
Comment (polaroi8d) [inline cli/runner.go:262]: I am just curious what is delnet?

**Review by polaroi8d — CHANGES_REQUESTED** [review]
Please add a go linter in your local config.

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go]: fixed

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go:250]: Network deletion if it exists with wrong parameters like wrong driver. Now it defaults to true, planned to be changeable.

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go]: fixed

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go]: fixed

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go]: fixed

File: cli/dyo.go
Comment (c3ppc3pp) [reply cli/dyo.go:105]: cCtx is the de facto for cli ConTeXt, where the cli is the library we're using.

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go]: Matter of taste, but yes.

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go]: fixed

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go:196]: Whitespace will be inserted after every parameter using log.Println()

File: cli/containergen.go
Comment (c3ppc3pp) [reply cli/containergen.go]: fixed

File: cli/containergen.go
Comment (c3ppc3pp) [reply cli/containergen.go]: fixed

File: cli/containergen.go
Comment (c3ppc3pp) [reply cli/containergen.go]: removed

File: cli/containergen.go
Comment (c3ppc3pp) [reply cli/containergen.go]: fixed

File: cli/dyo.go
Comment (c3ppc3pp) [reply cli/dyo.go:46]: removed

File: cli/dyo.go
Comment (c3ppc3pp) [reply cli/dyo.go:55]: removed

File: cli/dyo.go
Comment (c3ppc3pp) [reply cli/dyo.go:93]: fixed

File: web/crux/install-docker.sh.hbr
Comment (c3ppc3pp) [inline web/crux/install-docker.sh.hbr:42]: For windows MSYS installs (like git-bash) support.

File: cli/dyo.go
Comment (c3ppc3pp) [reply cli/dyo.go:1]: I'm open to suggestions, as this file is mostly serves as an entry point.

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go:195]: fixed

File: cli/dyo.go
Comment (c3ppc3pp) [reply cli/dyo.go]: fixed

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go]: changed

File: cli/containergen.go
Comment (c3ppc3pp) [reply cli/containergen.go]: fixed

File: agent/internal/grpc/grpc.go
Comment (c3ppc3pp) [reply agent/internal/grpc/grpc.go:417]: @m8vago can I remove it?

Comment (m8vago) [reply agent/internal/grpc/grpc.go:417]: yap, sure

File: cli/dyo.go
Comment (robot9706) [reply cli/dyo.go:1]: `main.go` or `entry.go` ?

File: cli/dyo.go
Comment (robot9706) [inline cli/dyo.go:15]: This evaluates at runtime, so Compiled will be now every time the CLI is run :)

**Review by robot9706 — CHANGES_REQUESTED** [review]

File: cli/dyo.go
Comment (c3ppc3pp) [reply cli/dyo.go]: fixed

File: cli/default_utils.go
Comment (m8vago) [inline cli/default_utils.go:13]: what's this burned in pw?

File: cli/default_crux.go
Comment (m8vago) [inline cli/default_crux.go:45]: magic pw

File: cli/default_utils.go
Comment (c3ppc3pp) [reply cli/default_utils.go:13]: predefined value for now

File: cli/default_crux.go
Comment (c3ppc3pp) [reply cli/default_crux.go:13]: predefined value, crux's postgres is using this one

File: cli/config_file.go
Comment (c3ppc3pp) [reply cli/config_file.go:27]: not yet, plan is using it as a one-shot command without other inputs, detailed settings is somewhat blocked by other factors right now

File: cli/main.go
Comment (polaroi8d) [inline cli/main.go:15]: dyrector.io

File: cli/main.go
Comment (polaroi8d) [inline cli/main.go:20]: I think use one word word one command its better, but talk about this in person.

File: cli/main.go
Comment (polaroi8d) [inline cli/main.go:71]: writeCompose?

File: cli/container_gen.go
Comment (polaroi8d) [inline cli/container_gen.go:119]: Needed to outsource in a function?

File: cli/dyo.go
Comment (polaroi8d) [reply cli/dyo.go]: They are using ctx, or c look https://pkg.go.dev/github.com/urfave/cli#ShowCommandHelp or am I misunderstand something?

File: cli/main.go
Comment (nandor-magyar) [reply cli/main.go:20]: As per synonyms, "up" would be nice too, but we don't need so many. Generate is more appropriate, if there is i why there is no install pair for it? We can have it consistent with gen only.

up: could do gen and up in sequence, in another command.

File: cli/runner.go
Comment (nandor-magyar) [inline cli/runner.go:15]: The thing that is running is not the cli (duh), it would look better using the product's name or similar: `dyrectorio-stack`, `dyrectorio-docker` or dev suffix.

File: cli/runner.go
Comment (nandor-magyar) [inline cli/runner.go:114]: Error doesn't contain enough info about the executable we are looking for, bad error message, insert more info.

File: cli/template.hbr
Comment (nandor-magyar) [inline cli/template.hbr:1]: This file is not a handlebars template, or is it?

File: cli/default_crux.go
Comment (c3ppc3pp) [reply cli/default_crux.go:45]: will be fixed in the next iteration

File: cli/container_gen.go
Comment (c3ppc3pp) [reply cli/container_gen.go:119]: Was used by multiple functions, and was more complicated. It will be reworked in the next iteration.

File: cli/main.go
Comment (c3ppc3pp) [reply cli/main.go]: fixed

File: cli/main.go
Comment (c3ppc3pp) [reply cli/main.go]: we talked about it, fixed

File: cli/container_gen.go
Comment (c3ppc3pp) [reply cli/container_gen.go]: fixed

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go]: done

File: cli/template.hbr
Comment (c3ppc3pp) [reply cli/template.hbr]: it is not, renamed

File: cli/runner.go
Comment (c3ppc3pp) [reply cli/runner.go]: changed to dyrectorio-stack

File: cli/main.go
Comment (c3ppc3pp) [reply cli/main.go]: good catch, thanks!

File: cli/dyo.go
Comment (c3ppc3pp) [reply cli/dyo.go]: Internally, yes. Please look here https://cli.urfave.org/v2/examples/flags/

File: cli/runner.go
Comment (robot9706) [inline cli/runner.go:71]: 🤣

**Review by robot9706 — APPROVED** [review]
😎

**Comment by szilarddoro** [comment]
https://docs.dyrector.io/get-started/cli

Your docs mention `go run . start` which is inaccurate, because on the latest version I need to use the `go run . up` command to start the CLI.

**Review by nandor-magyar — COMMENTED** [review]
Waiting for the discussed refactor.

**Comment by polaroi8d** [comment]
> https://docs.dyrector.io/get-started/cli
> 
> Your docs mention `go run . start` which is inaccurate, because on the latest version I need to use the `go run . up` command to start the CLI.

Thanks, fixed.

File: .vscode/launch.json
Comment (polaroi8d) [inline .vscode/launch.json:41]: minor; but `CLI`

File: .vscode/launch.json
Comment (polaroi8d) [inline .vscode/launch.json:71]: same

File: cli/Makefile
Comment (polaroi8d) [inline cli/Makefile:1]: Why we call it "package" instead of "agent"?

File: cli/config_file.go
Comment (polaroi8d) [inline cli/config_file.go:91]: All words started with uppercase.

File: cli/config_file.go
Comment (polaroi8d) [inline cli/config_file.go:97]: I don't know which one would be better but prefer something like 'dyo-cli' or 'dyrector.io'

File: cli/config_file.go
Comment (polaroi8d) [inline cli/config_file.go:151]: Missing username from TODO.

File: cli/config_file.go
Comment (robot9706) [reply cli/config_file.go:97]: 'dyo-cli' would be fine I think 🤔

File: cli/config_file.go
Comment (c3ppc3pp) [reply cli/config_file.go]: Added.

File: cli/config_file.go
Comment (c3ppc3pp) [reply cli/config_file.go]: "dyo" as default settings directory was changed to dyo-cli.

File: cli/config_file.go
Comment (c3ppc3pp) [reply cli/config_file.go]: Fixed, thanks!

File: cli/Makefile
Comment (c3ppc3pp) [reply cli/Makefile]: Removed the unnecessary lines. The file will be extended or removed when CLI will get another update (go install -able, and added to the releases) depending on our preference of global makefile.

**Review by polaroi8d — APPROVED** [review]
LGTM (tested in locally)

File: .vscode/launch.json
Comment (c3ppc3pp) [reply .vscode/launch.json]: Fixed.

File: .vscode/launch.json
Comment (c3ppc3pp) [reply .vscode/launch.json]: Nice catch, CLI is an abbreviation. Fixed.

**Comment by polaroi8d** [comment]
After the merge, we have to update the root README.md about the local development.

File: cli/dyo.go
Comment (c3ppc3pp) [reply cli/dyo.go]: @polaroi8d Following example codes are sufficient, or should I rename the context variables?

Comment (polaroi8d) [reply cli/dyo.go]: Okey for me, thanks ✅

**Pull request merged**
